sensor/common/scan/scan.go

```diff
@@ -163,57 +131,171 @@ func (s *LocalScan) enrichLocalImageFromRegistry(ctx context.Context, centralCli
+	// errs are only added to errorList when attempts from all pull sources + all registries fail.
+	errs := errorhelpers.NewErrorList("")
+
+	// For each pull source, obtain the associated registries and attempt to obtain metadata, stopping on first success.
+	for _, pullSource := range pullSources {
+		registries, err := s.getRegistries(ctx, namespace, pullSource.GetName())
+		if err != nil {
+			log.Warnf("Error getting registries for pull source %q, skipping: %v", pullSource.GetName().GetFullName(), err)
+			errs.AddError(err)
+			continue
+		}
+
+		log.Debugf("Using %v registries for enriching pull source %q", len(registries), pullSource.GetName().GetFullName())
+
+		// Create an image and attempt to enrich it with metadata.
+		pullSourceImage := types.ToImage(pullSource)
+		reg := s.enrichImageWithMetadata(errs, registries, pullSourceImage)
+		if reg != nil {
+			// Successful enrichment.
+			enrichImageDataSource(sourceImage, reg, pullSourceImage)
+
+			srcName := sourceImage.GetName().GetFullName()
+			srcID := sourceImage.GetId()
+			pullName := pullSourceImage.GetName().GetFullName()
+			pullID := pullSourceImage.GetId()
+			log.Infof("Image %q (%v) enriched with metadata using pull source %q (%v) and registry %q", srcName, srcID, pullName, pullID, reg.Name())
+			log.Debugf("Metadata for image %q (%v) using pull source %q (%v): %v", srcName, srcID, pullName, pullID, pullSourceImage.GetMetadata())
+			return reg, pullSourceImage
+		}
+	}
+
+	// Attempts for every pull source and registry have failed.
+	errorList.AddErrors(errs.Errors()...)
+
+	image := types.ToImage(sourceImage)
+	image.Notes = append(image.Notes, storage.Image_MISSING_METADATA)
+	return nil, image
+}
+
+func enrichImageDataSource(sourceImage *storage.ContainerImage, reg registryTypes.ImageRegistry, targetImg *storage.Image) {
+	ds := reg.DataSource().Clone()
+
+	// If target is a mirror, then add mirror details to the data source.
+	if sourceImage.GetName().GetFullName() != targetImg.GetName().GetFullName() {
+		ds.Mirror = targetImg.GetName().GetFullName()
+		ds.MirrorClusterId = clusterIDGetter()
+	}
+
+	targetImg.GetMetadata().DataSource = ds
+}
+
+// getRegistries will return registries that match the provided image, starting with image integrations sync'd from Central,
+// then namespace pull secrets, and lastly any global pull secrets (such as the OCP global pull secret). If no registries
+// are found will return a new registry that has no credentials.
+func (s *LocalScan) getRegistries(ctx context.Context, namespace string, imgName *storage.ImageName) ([]registryTypes.ImageRegistry, error) {
+	var regs []registryTypes.ImageRegistry
+
+	// Add registries from Central's image integrations.
+	centralIntegrations := s.getMatchingCentralRegIntegrations(imgName)
+	if len(centralIntegrations) > 0 {
+		regs = append(regs, centralIntegrations...)
+	}
+
+	// Add registries from k8s pull secrets.
+	if namespace != "" {
+		// If namespace provided pull appropriate registry.
```

Comment:
"pull appropriate" reminds me. We have a concept of knowing if an image is "pullable". Does this come into play when dealing with mirror registries?

Reply:
Yes, in a way - prior to an image being sent for 'scanning', Sensor's enricher will check if the image was marked as not pullable [here](https://github.com/stackrox/stackrox/blob/919d8f0c9b3c04c53ab5e436044e34c854d9e666/sensor/common/detector/enricher.go#L199).

If it is pullable, then its reasonable to assume that at least one mirror or the source image is pullable - because the runtime was able to pull it.